Add tests to styleguide & fix pending job live duration

rupurt opened the pull request:
- TS typechecking support during tests
- Extract `<ElapsedDuration />` with tests from OperatorUI

![Screen Shot 2020-01-14 at 1 13 12 PM](https://user-images.githubusercontent.com/680789/72383147-b28df780-36cf-11ea-8772-602dc090a004.png)

NavyAdmiral commented on styleguide/package.json:
```diff
@@ -38,14 +40,19 @@
     "@storybook/addons": "^5.2.1",
     "@storybook/react": "^5.1.10",
+    "@testing-library/jest-dom": "^4.2.4",
+    "@testing-library/react": "^9.4.0",
```
:heart:

HenryNguyen5 commented on styleguide/src/components/ElapsedDuration.tsx:
```diff
@@ -0,0 +1,17 @@
+
+export const ElapsedDuration: React.FC<Props> = ({ start, end, className }) => {
+  return (
+    <Typography role="elapsedduration" variant="h6" className={className}>
```
Should this be camel case?

rupurt replied:
My understanding is that ARIA roles should be lowercase combined words. e.g. https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Rowgroup_Role

I switched it to a role from an id so that there can be multiple on the same page FWIW.

HenryNguyen5 commented on .circleci/config.yml:
```diff
@@ -140,9 +156,7 @@ jobs:
-      - run: yarn workspace @chainlink/json-api-client
-      - store_artifacts:
-          path: ./integration/logs
+      - run: yarn workspace @chainlink/json-api-client run test
```
Tip: you can omit `run` and it'll still work :D

rupurt replied:
👍 because it's CI config I prefer it to be the explicit full command

HenryNguyen5 replied:
sounds good

HenryNguyen5 commented on styleguide/jest.config.js:
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+  preset: 'ts-jest/presets/js-with-ts',
```
Seems like this would work fine, since we don't have any js in styleguide
```suggestion
  preset: 'ts-jest',
```

HenryNguyen5 commented on styleguide/jest.config.js:
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+  preset: 'ts-jest/presets/js-with-ts',
+  moduleDirectories: ['node_modules', '<rootDir>/src/'],
```
Do we need this? I think we can drop it

HenryNguyen5 commented on styleguide/jest.config.js:
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+  preset: 'ts-jest/presets/js-with-ts',
+  moduleDirectories: ['node_modules', '<rootDir>/src/'],
+  testPathIgnorePatterns: ['<rootDir>/dist/', '<rootDir>/tmp/'],
```
When does the `tmp` dir become created? Haven't seen it day-to-day

rupurt replied:
I think it's a carry over from the webpack dev server in OperatorUI. ✂️

HenryNguyen5 commented on styleguide/package.json:
```diff
@@ -12,7 +12,9 @@
-    "setup": "yarn build"
+    "setup": "yarn build",
+    "test": "jest",
+    "test:ci": "yarn test --coverage --reporters jest-silent-reporter --maxWorkers=50%"
```
Out of curiosity, why we do use the silent reporter?

rupurt replied:
So that there's less output when everything passes. I don't feel strongly about it.

HenryNguyen5 replied:
Maybe we should just remove the dep across the project, one less dep to worry about

HenryNguyen5 commented on styleguide/src/utils/elapsedDuration.ts:
```diff
@@ -0,0 +1,26 @@
+const SECONDS_IN_HOUR = 3600
+
+export function elapsedDuration(createdAt: string, finishedAt: string): string {
+  if (!createdAt && !finishedAt) {
```
Should we make this more explicit? Something like
```suggestion
  if (createdAt === '' && finishedAt === '') {
```

HenryNguyen5 commented on styleguide/src/utils/elapsedDuration.ts:
```diff
@@ -0,0 +1,26 @@
+const SECONDS_IN_MINUTE = 60
+const SECONDS_IN_HOUR = 3600
+
+export function elapsedDuration(createdAt: string, finishedAt: string): string {
```
Can we have a jsdoc here explaining what format the strings in the `createdAt` and `finishedAt` parameters should be? And what format the string returned by `elapsedDuration` is?

rupurt replied:
Sorry missed this comment. Added in https://github.com/smartcontractkit/chainlink/pull/2140/commits/2a38af210978ee9b4ac2dbccba757631360a8a2c. Is that along the lines of what you were thinking?

HenryNguyen5 commented on styleguide/src/components/ElapsedDuration.tsx:
```diff
@@ -0,0 +1,17 @@
+
+export const ElapsedDuration: React.FC<Props> = ({ start, end, className }) => {
+  return (
+    <Typography role="elapsedduration" variant="h6" className={className}>
```
FWIW, if its only for testing, we should be using stable `test-ids` for our components

rupurt replied:
This was only used for testing. I didn't explore how it was used when I originally did this. After poking around I noticed that the duration for pending jobs is busted on `develop`. Fixed the issue and removed the role in https://github.com/smartcontractkit/chainlink/pull/2140/commits/f8bb654e0b742949de089872aa3639fa6b4e2e81

rupurt commented on styleguide/src/components/ElapsedDuration.tsx:
```diff
@@ -0,0 +1,17 @@
+}
+
+export const ElapsedDuration: React.FC<Props> = ({ start, end, className }) => {
+  return (
```
I've found them frustrating as they need to be converted back to this form when you need to use hooks.

How do you feel about implicit returns here?

rupurt commented:
@HenryNguyen5 @hellobart addressed the comments. I changed several things and added extra test coverage and type annotations to cover the cases discovered in the bug mentioned above. Probably worth doing a re-review of the whole thing.

HenryNguyen5 commented on styleguide/__tests__/components/ElapsedDuration.test.tsx:
```diff
@@ -0,0 +1,15 @@
+const END = '2020-01-03T23:45:30.166261Z'
+
+describe('ElapsedDuration', () => {
+  test('displays the duration between 2 dates in a human readable foramt', () => {
```
Can we continue using `it` here?

hellobart replied:
displays the duration between 2 dates in a human readable **foramt** <- typo

HenryNguyen5 commented on styleguide/__tests__/utils/elapsedDuration.test.ts:
```diff
@@ -0,0 +1,47 @@
+const START = '2020-01-03T22:45:00.166261Z'
+
+describe('elapsedDuration', () => {
+  test('only displays seconds when < 1 min', () => {
```
Can we continue using `it` here?

HenryNguyen5 commented on styleguide/src/utils/elapsedDuration.ts:
```diff
@@ -0,0 +1,38 @@
+export type DateValue = string | number
+export type FinishedAt = DateValue | null
+
+export function elapsedDuration(
```
Could we use a comment here? Expected format and expected return value?
